### PR TITLE
Recurrent changes

### PR DESCRIPTION
**Pull request by AdrienClairembault**
Add recurrent changes.
Diff is not as big as it looks, most of it is code moving from TicketRecurrent to the new CommonITILRecurrent parent.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Internal ref | !20911

### CODE REVIEW

**Comment by cedric-anne** [comment]
We have some issues on edge cases on recurrent tickets creation that appears when tests are triggered at specific hours, depending on `create_before`.
Issue is due to following facts :
 - computation is always done from `begin_date`,
 - `next_creation_date` corresponds to the ticket occurence date altered by `create_before` delay.

A solution to fix this issue could be to :
 - store the last ticket occurence date in `last_occurence_date` (the real ticket occurence date, not altered by `create_before`),
 - replace `next_creation_date` by `next_occurence_date` (to contains the real ticket occurence date, not altered by `create_before`),
 - adapt next occurence computation to start from the last occurence date,
 - adapt SQL query of the cron to create tickets when ``UNIX_TIMESTAMP(next_occurence_date) - create_before <  UNIX_TIMESTAMP(NOW()) ``

Do you think it should be done on this PR, or in another PR ?

**Comment by AdrienClairembault** [comment]
> 
> 
> We have some issues on edge cases on recurrent tickets creation that appears when tests are triggered at specific hours, depending on `create_before`.
> Issue is due to following facts :
> 
>     * computation is always done from `begin_date`,
> 
>     * `next_creation_date` corresponds to the ticket occurence date altered by `create_before` delay.
> 
> 
> A solution to fix this issue could be to :
> 
>     * store the last ticket occurence date in `last_occurence_date` (the real ticket occurence date, not altered by `create_before`),
> 
>     * replace `next_creation_date` by `next_occurence_date` (to contains the real ticket occurence date, not altered by `create_before`),
> 
>     * adapt next occurence computation to start from the last occurence date,
> 
>     * adapt SQL query of the cron to create tickets when `UNIX_TIMESTAMP(next_occurence_date) - create_before <  UNIX_TIMESTAMP(NOW()) `
> 
> 
> Do you think it should be done on this PR, or in another PR ?

Another PR, I prefer to not overload the backport for the internal ref.

**Comment by trasher** [comment]
> Do you think it should be done on this PR, or in another PR ?
Well, that should be fixed on 9.5 if possible, so another PR. I guess we first need a test to reproduce that issue; to be sure it's fixed.

File: inc/commonitilrecurrentcron.class.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+               ]
+            ]
+         ]);
+
+         while ($data = $iterator->next()) {
+            /** @var CommonITILRecurrent */
+            $item = new $itemtype();
+            $item->fields = $data;
+
+            if ($item->createItem($data)) {
```
Comment (cedric-anne) [inline inc/commonitilrecurrentcron.class.php:87]: $data seems not used anymore
```suggestion
            if ($item->createItem()) {
```

File: inc/commonitilrecurrentcron.class.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/**
+ * Cron task creating recurrent tickets and changes
+ *
+ * @since NEXT_MAJOR_VERSION
```
Comment (cedric-anne) [inline inc/commonitilrecurrentcron.class.php:36]:
```suggestion
 * @since x.x.x
```

File: install/update_95_xx.php [hunk 3]
```diff
@@ -94,6 +94,47 @@ function update95toXX() {
...
+         KEY `changetemplates_id` (`changetemplates_id`),
+         KEY `next_creation_date` (`next_creation_date`)
+      ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;");
+   }
+
+   $migration->addRight(RecurrentChange::$rightname, ALLSTANDARDRIGHT, [
+      Change::$rightname => UPDATE
+   ]);
+   // /Recurrent changes
+
```
Comment (cedric-anne) [inline install/update_95_xx.php:137]: Please put this in a distinct file (in `install/update_95_xx` directory).

File: install/update_95_xx.php [hunk 4]
```diff
@@ -94,6 +94,47 @@ function update95toXX() {
       }
    }
 
+   // Recurrent changes
+   $DB->updateOrDie(CronTask::getTable(), [
+         'itemtype' => CommonITILRecurrentCron::class,
```
Comment (cedric-anne) [inline install/update_95_xx.php:99]: I think using hard coded strings is better in migration, to prevent migration to be altered if we refactor the code.
```suggestion
         'itemtype' => 'CommonITILRecurrentCron',
```

File: install/update_95_xx.php [hunk 5]
```diff
@@ -94,6 +94,47 @@ function update95toXX() {
...
+         KEY `is_recursive` (`is_recursive`),
+         KEY `is_active` (`is_active`),
+         KEY `changetemplates_id` (`changetemplates_id`),
+         KEY `next_creation_date` (`next_creation_date`)
+      ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;");
+   }
+
+   $migration->addRight(RecurrentChange::$rightname, ALLSTANDARDRIGHT, [
+      Change::$rightname => UPDATE
+   ]);
```
Comment (cedric-anne) [inline install/update_95_xx.php:135]: Same remark here.

File: inc/commonitilrecurrent.class.php [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+/**
+ * Base class for recurrent tickets and changes
+ *
+ * @since NEXT_MAJOR_VERSION
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/commonitilrecurrent.class.php:36]: Well... Would need to be changed on all existing files aswell; maybe doing at once is another PR would be better

**Review by cedric-anne — APPROVED** [review]

**Comment by trasher** [comment]
Please rebase to solve conflict. We've made several changes in tests that have just been merged, so it is not impossible there new fails after rebase.

File: install/update_95_xx.php [hunk 7]
```diff
@@ -49,6 +49,51 @@ function update95toXX() {
    require __DIR__ . '/update_95_xx/domains.php';
    require __DIR__ . '/update_95_xx/devicebattery.php';
    require __DIR__ . '/update_95_xx/reservationitem.php';
+   require __DIR__ . '/update_95_xx/recurrentchange.php';
+
+   //add is_fqdn on some domain records types
```
Comment (trasher) [inline install/update_95_xx.php:54]: This should probably not be here

Comment (AdrienClairembault) [reply install/update_95_xx.php]: Indeed, error during rebase.
I've rebased again to fix conflicts, should be good now.

File: install/update_95_xx.php [hunk 8]
```diff
@@ -50,6 +50,50 @@ function update95toXX() {
...
+      }
+
+      if ($updated) {
+         $DB->update(
+            DomainRecordType::getTable(),
+            ['fields' => json_encode($type_fields)],
+            ['name' => $field['name']]
+         );
+      }
+   }
```
Comment (trasher) [inline install/update_95_xx.php:96]: still not ok

Comment (AdrienClairembault) [reply install/update_95_xx.php:96]: Just saw it too, fixing asap

Comment (AdrienClairembault) [reply install/update_95_xx.php]: All good now :)

**Review by trasher — APPROVED** [review]
